File: P1/Ejercicio1/factoria/FactoriaCarreraYBicicleta.java
package factoria;

import bicicleta.Bicicleta;
import carrera.Carrera;

/**
 * Interfaz para representar la factoría que se encargará de crear las carreras y las bicicletas
 */
public interface FactoriaCarreraYBicicleta {
    /**
     * Crea una carrera de bicicletas
     * @param numBicicletas Número de bicicletas que contendrá la carrera
     * @return Objeto Carrera con el tipo de carrera de bicicletas creada
     */
    Carrera crearCarrera(int numBicicletas);

    /**
     * Crea una bicicleta
     * @param id Identificador que tendrá la bicicleta
     * @return Objeto Bicicleta con el tipo de bicicleta creada
     */
    Bicicleta crearBicicleta(int id);
}

File: P1/Ejercicio2/Factorias/factoriaCarreraYBicicleta.py
from abc import ABC, abstractmethod
from Bicicletas.bicicleta import Bicicleta

class FactoriaCarreraYBicicleta(ABC):
    """
    Clase abstracta que define la interfaz para las fábricas de carreras y bicicletas.
    Establece los métodos para crear carreras y bicicletas específicas según el tipo.
    
    Atributos:
      prototipo_bicicleta: la bicicleta prototipo que se copiará y se crearán el resto de bicicletas.
    """
    
    def __init__(self, bicicleta: Bicicleta):
        """
        Inicialización del prototipo de bicicleta para clonación.
        """
        self._prototipo_bicicleta = bicicleta
        
    @abstractmethod
    def crear_item(self, item, id):
        """
        Método abstracto para crear un item específico, ya sea una carrera o una bicicleta de montaña o carretera.
        
        Parametros:
            item: cadena de texto que describe el objeto a crear ('carrera' o 'bicicleta').
            id: identificador númerico para cada bicicleta.
        """
        pass